inc/admin/attachments/namespace.php

```diff
@@ -0,0 +1,172 @@
+<?php
+/**
+ * @author Brad Payne, Alex Paredes
+ * @license  GPLv3 (or any later version)
+ */
+
+namespace Pressbooks\Admin\Attachments;
+
+use Pressbooks\Licensing;
+use Pressbooks\Utility;
+
+/**
+ * Hooks into 'attachment_fields_to_edit' filter to add custom attachment
+ * metadata. Currently, on the display side of capturing this metadata
+ * unlike images, there is no easy way to determine an audio or video post id
+ * from the markup that's generated: (wp-image-33), therefore this can only
+ * apply to images, for now
+ *
+ * @since 5.5.0
+ *
+ * @param $form_fields
+ * @param $post
+ *
+ * @return mixed
+ */
+function add_metadata_attachment( $form_fields, $post ) {
+
+	if ( substr( $post->post_mime_type, 0, 5 ) === 'image' ) {
+
+		$author      = get_post_meta( $post->ID, 'pb_media_attribution_author', true );
+		$author_url  = get_post_meta( $post->ID, 'pb_media_attribution_author_url', true );
+		$source      = get_post_meta( $post->ID, 'pb_media_attribution_title_url', true );
+		$license     = get_post_meta( $post->ID, 'pb_media_attribution_license', true );
+		$figure      = get_post_meta( $post->ID, 'pb_media_attribution_figure', true );
+		$adapted     = get_post_meta( $post->ID, 'pb_media_attribution_adapted', true );
+		$adapted_url = get_post_meta( $post->ID, 'pb_media_attribution_adapted_url', true );
+
+		$form_fields['pb_attribution'] = [
+			'value' => '',
+			'label' => __( 'ATTRIBUTIONS', 'pressbooks' ),
+			'input' => 'html',
+			'html'  => '<span></span>',
+		];
+
+		$form_fields['pb_media_attribution_figure'] = [
+			'value' => isset( $figure ) ? $figure : '',
+			'label' => __( 'Figure #', 'pressbooks' ),
+			'input' => 'text',
+		];
+
+		$form_fields['pb_media_attribution_title_url'] = [
+			'value' => isset( $source ) ? $source : '',
+			'label' => __( 'Source', 'pressbooks' ),
+			'input' => 'html',
+			'html'  => "<input type='url' class='text urlfield' placeholder='https://creativecommons.org/' name='attachments[$post->ID][pb_media_attribution_title_url]' value='" . esc_attr( $source ) . "' />",
```

Comment:
I have mixed feelings about the placeholders. When I first opened a media attachment to edit, I thought for a split second that default information had been inserted. I think it might be best to omit placeholders for the URL fields unless the other fields will have them too, and rely on the field labels for clarity (as recommended at [this Mozilla Developer article](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#Labels_and_placeholders)).

Reply:
Hmm, yes so the reason behind the placeholder text was to provide an idea of what format the field expected for the URL. If it's not entered in a specific way it doesn't get saved, and there is no error message. Any suggestions on how to improve that, or would you still prefer this placeholder gone?

Reply:
Looking at the filter documentation for `attachment_fields_to_edit`, I found a link to this tutorial which suggests that validation errors can be added and displayed via the `attachment_fields_to_save` hook: https://code.tutsplus.com/articles/creating-custom-fields-for-attachments-in-wordpress--net-13076

I think that would be my preferred approach. It looks like you can also add an array key called `'helps'` to each field with descriptive text which would appear below the field, which might be a good way to provide additional context prior to entry and validation.

Reply:
thanks for that quick feedback, we couldn't get the validation errors to work, but I will try again. Failing that, will use the descriptive text as suggested.